From cf7f29ffc336757dd683f9f4312ebc6f7e546654 Mon Sep 17 00:00:00 2001
From: Dennis Cagle <dcagle@codeaurora.org>
Date: Fri, 17 Sep 2021 14:42:25 -0700
Subject: [PATCH] device-mapper: Poll for mmcblk0p33 availablility

MMC bus performs device scanning in the background which causes
the device-mapper early init driver to not see required devices.
Wait for device to actually be available before continuing.

Change-Id: Ic3a68bca2e385532db6ab6af1475499d5c1e87c8
Signed-off-by: Dennis Cagle <dcagle@codeaurora.org>
---
 drivers/md/dm-init.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/drivers/md/dm-init.c b/drivers/md/dm-init.c
index b869316..47e05d8 100644
--- a/drivers/md/dm-init.c
+++ b/drivers/md/dm-init.c
@@ -8,11 +8,13 @@
  */
 
 #include <linux/ctype.h>
+#include <linux/delay.h>
 #include <linux/device.h>
 #include <linux/device-mapper.h>
 #include <linux/init.h>
 #include <linux/list.h>
 #include <linux/moduleparam.h>
+#include <linux/mount.h>
 
 #define DM_MSG_PREFIX "init"
 #define DM_MAX_DEVICES 256
@@ -267,6 +269,7 @@ static int __init dm_init_init(void)
 	LIST_HEAD(devices);
 	char *str;
 	int r;
+	dev_t bdev;
 
 	if (!create)
 		return 0;
@@ -286,6 +289,15 @@ static int __init dm_init_init(void)
 	DMINFO("waiting for all devices to be available before creating mapped devices");
 	wait_for_device_probe();
 
+	// Poll for block device availablility
+	bdev = name_to_dev_t("/dev/mmcblk0p33");
+	DMINFO("waiting for block device availability");
+	while (!bdev) {
+		msleep(100);
+		bdev = name_to_dev_t("/dev/mmcblk0p33");
+		DMINFO("block device not available, retrying");
+	}
+
 	list_for_each_entry(dev, &devices, list) {
 		if (dm_early_create(&dev->dmi, dev->table,
 				    dev->target_args_array))
-- 
2.7.4

